### PR TITLE
chores

### LINKED ISSUE #802: Warning comparison of integer expressions of different signedness
```
./thirdparty/manifold/src/manifold/include/manifold.h: In member function 'uint32_t manifold::MeshGL::NumVert() const':
Error: ./thirdparty/manifold/src/manifold/include/manifold.h:50:41: error: comparison of integer expressions of different signedness: 'std::vector<float>::size_type' {aka 'long unsigned int'} and 'int' [-Werror=sign-compare]
   50 |     if (vertProperties.size() / numProp >= std::numeric_limits<int>::max())
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./thirdparty/manifold/src/manifold/include/manifold.h: In member function 'uint32_t manifold::MeshGL::NumTri() const':
Error: ./thirdparty/manifold/src/manifold/include/manifold.h:56:41: error: comparison of integer expressions of different signedness: 'std::vector<float>::size_type' {aka 'long unsigned int'} and 'int' [-Werror=sign-compare]
   56 |     if (vertProperties.size() / numProp >= std::numeric_limits<int>::max())
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by pca006132**
Fix #802, #804 and https://github.com/elalish/manifold/pull/832#issuecomment-2155831898

@elalish When doing this change, I found that we are using 32-bit integers extensively. I feel like the proper way would be to use 64-bit integers. There may be some performance hit, but I doubt if it will be huge. And it just feels better if we can make things work with larger meshes, considering we don't have any reasons not to do so? Some public APIs will have to be changed, but they will be changed when we update to double precision floating point.

For #804, it should now build fine, but it is not the best way of handling exceptions. We need more careful checks everywhere, probably together with #744.

### CODE REVIEW

**Comment by pca006132** [comment]
probably messed up the python binding somewhere, will fix tmr.

File: src/manifold/include/manifold.h
Comment (elalish) [inline src/manifold/include/manifold.h:63]: I think I don't understand something about C++ exceptions. I thought any `throw` just automatically became `exit` when exceptions are disabled in the compiler. Or does disabling exceptions just turn uses of `throw` into compile errors?

File: src/utilities/include/vec_view.h
Comment (elalish) [inline src/utilities/include/vec_view.h:60]: Should we make another wrapper for these a bit like our `ASSERT`?

**Review by elalish — COMMENTED** [review]
Thanks, this looks great. Given the breaking changes, what do you say we hold this for ~week while we release v2.5, then put this in for v3.0 with the switch to doubles and all?

File: src/utilities/include/vec_view.h
Comment (pca006132) [reply src/utilities/include/vec_view.h:60]: probably, my goal for now is to get it to compile
I think we need better error handling

File: src/manifold/include/manifold.h
Comment (pca006132) [reply src/manifold/include/manifold.h:63]: I vaguely recall that emscripten requires no `throw` statement to compile with exception disabled. Not really sure about that.

Comment (elalish) [reply src/manifold/include/manifold.h:63]: Okay, but how about in normal C++ workflows? I still don't really understand #804. Is there something we did that made it impossible to build with exceptions disabled? @fire - can you help me understand https://github.com/godotengine/godot-proposals/issues/9711#issuecomment-2146063161? Our exceptions at this point are only a convenience for debug and such. Is there something that breaks when you compile with exceptions disabled?

Comment (fire) [reply src/manifold/include/manifold.h:63]: When we compile with exceptions enabled, for example web, and a few of the platforms break. We couldn't disable exceptions.

Comment (elalish) [reply src/manifold/include/manifold.h:63]: Ah, so this is specific to web. In that case this PR may well address the issue. We could really use this kind of detail here or in #804. Which platforms break? What were the errors? Why couldn't you disable exceptions? Do you want to disable exceptions? 

I really want to help you out, but I do need at least *some* feedback.

Comment (fire) [reply src/manifold/include/manifold.h:63]: Here's what I can do. I'll push the branch with exceptions on and watch godot engine cid crash. Will post the branch.

Comment (fire) [reply src/manifold/include/manifold.h:63]: https://github.com/V-Sekai/godot/tree/vsk-csg-exceptions-4.3

**Comment by pca006132** [comment]
Converting everything to size_t turns out to be quite complicated, mainly due to `SparseIndices` and `Collider`. Probably need some more time to do it, and evaluate the performance differences. This can be merged even though the changes are incomplete.

**Comment by pca006132** [comment]
@elalish note that the old `ASSERT` is now `DEBUG_ASSERT`, and `ASSERT` is for things that should still be checked when `MANIFOLD_DEBUG` is disabled.

File: src/manifold/src/boolean3.cpp
Comment (elalish) [inline src/manifold/src/boolean3.cpp:142]: what is `ssize_t`? Okay, I found [this](https://stackoverflow.com/questions/19224655/using-ssize-t-vs-int) and am slightly more confused.

File: src/manifold/src/properties.cpp
Comment (elalish) [inline src/manifold/src/properties.cpp:204]: Will the update of `Halfedge` happen with `Collider`?

File: src/manifold/src/smoothing.cpp
Comment (elalish) [inline src/manifold/src/smoothing.cpp:437]: Why not the same as the change just below?

File: src/manifold/src/subdivision.cpp
Comment (elalish) [inline src/manifold/src/subdivision.cpp:130]: 👍

File: src/polygon/src/polygon.cpp
Comment (elalish) [inline src/polygon/src/polygon.cpp:953]: 👍

File: src/utilities/include/utils.h
Comment (elalish) [inline src/utilities/include/utils.h:23]: This makes me slightly worried about `ssize_t`...

File: src/utilities/include/vec.h
Comment (elalish) [inline src/utilities/include/vec.h:18]: Do we need to include this per file, or can it just be wherever we define `ASSERT`?

File: src/utilities/include/vec_view.h
Comment (elalish) [inline src/utilities/include/vec_view.h:19]: same as above.

File: .github/workflows/manifold.yml
Comment (elalish) [inline .github/workflows/manifold.yml:112]: Is there any reason to build WASM with exceptions off? Maybe better to test one of the nix builds since they're fast?

Comment (fire) [reply .github/workflows/manifold.yml:112]: I have a use for wasm with exceptions off.

Comment (elalish) [reply .github/workflows/manifold.yml:112]: @fire I think your earlier comment mentioned you do build WASM with exceptions off - that requires special CMake instructions which I was considering removing since JS is generally exceptions-based. But if WASM sans exceptions is actually useful to someone, then @pca006132 has it set up properly now. Do you build for WASM without exceptions?

Comment (fire) [reply .github/workflows/manifold.yml:112]: I would expect to build Godot Engine WASM without exceptions since the Godot Engine editor and Godot Engine game templates build without exceptions in the default configuration on all platforms.

Comment (elalish) [reply .github/workflows/manifold.yml:112]: Thanks! Okay, let's leave it as is in this PR then.

File: bindings/python/manifold3d.cpp
Comment (pca006132) [reply bindings/python/manifold3d.cpp:71]: Ah, should fix this. Thanks.

File: src/cross_section/src/cross_section.cpp
Comment (pca006132) [reply src/cross_section/src/cross_section.cpp:190]: No, this is also backward, `rbegin` and `rend` are reverse iterators.

File: src/manifold/include/manifold.h
Comment (pca006132) [reply src/manifold/include/manifold.h:51]: I forgot why, need to check. Probably due to `MeshGL` indices are/should be unsigned integers?

File: src/manifold/src/boolean3.cpp
Comment (pca006132) [reply src/manifold/src/boolean3.cpp:142]: `ssize_t` is signed version of `size_t`. But thinking about it, we only need that -1 to signal failure, and I can do it with `size_t` max as well.

File: src/manifold/src/edge_op.cpp
Comment (pca006132) [reply src/manifold/src/edge_op.cpp:159]: Will fix.

File: src/manifold/src/properties.cpp
Comment (pca006132) [reply src/manifold/src/properties.cpp:204]: Yes.

File: src/manifold/src/smoothing.cpp
Comment (pca006132) [reply src/manifold/src/smoothing.cpp:437]: Probably some error when I was too sleepy writing this, will fix.

File: src/utilities/include/utils.h
Comment (pca006132) [reply src/utilities/include/utils.h:23]: yeah, thinking about it `ssize_t` is not really needed.

File: src/utilities/include/vec.h
Comment (pca006132) [reply src/utilities/include/vec.h:18]: yeah probably just need it when we define `ASSERT`

File: src/cross_section/src/cross_section.cpp
Comment (elalish) [reply src/cross_section/src/cross_section.cpp:190]: ah, reading too quickly, thanks.

**Comment by pca006132** [comment]
sorry, was having a headache today. Will update this tmr

**Comment by pca006132** [comment]
I also updated nanobind version.

**Comment by pca006132** [comment]
there are still a lot of uses of integer that should be fixed, but we should probably leave to another PR and unblock your iterator changes.

**Review by elalish — APPROVED** [review]
Thanks, this looks great!

**Pull request merged**
